### PR TITLE
Fix 3806 testshell_markdown_tutorial_crypto: fails on macOS

### LINKED ISSUE #3806: testshell_markdown_tutorial_crypto: fails on macOS
## Steps to Reproduce the Problem

Fails on all macOS CI runners (Cirrus-CI, GitHub Actions). Notably GnuPG has been updated to 2.3.x and libgcrypt to 1.9.3.
See an [example](https://cirrus-ci.com/task/5540780325470208) run.

I was not yet able to reproduce the issue locally.

## Expected Result

Test should succeed.

## Actual Result

Truncated log:

```
1/1 Test #185: testshell_markdown_tutorial_crypto ...***Failed    4.75 sec

CMD: kdb mount test.ini user:/tests crypto "crypto/key=$(kdb gen-gpg-testkey)" base64 ini
RET: 7
=== FAILED return value does not match expected pattern 0
STDERR: The command /private/var/folders/sr/b58hwhtj0jbcf4r09zmg0wlc0000gn/T/cirrus-ci-build/build/bin/kdb-full mount terminated unsuccessfully with the info:
The provided plugin configuration is not valid!
Errors/Warnings during the check were:
Sorry, module crypto issued the error C01330:
Plugin Misbehavior: GPG failed with return value 512. gpg: (null): skipped: No public key
gpg: [stdin]: encryption failed: No public key

Mountpoint: /
Configfile: 
At: ../src/plugins/crypto/gpg.c:788

Please report the issue at https://issues.libelektra.org/
ERROR: C01330
```

## System Information

- Elektra Version: master
- Operating System: macOS
- Versions of other relevant software: GnuPG 2.3.1, libgcrypt 1.9.3

## Further Log Files and Output
Full test log: [tests.log.txt](https://github.com/ElektraInitiative/libelektra/files/6363698/tests.log.txt)

## Tasks

- [ ] fix issue? (or wait for upstream fix)
- [ ] use updated GnuPG series 2.3.x (undo **parts** of #3809)
- [ ] implement missing error handling of `gpgme_op_createkey` in tool `gen-gpg-testkey`

### PR DESCRIPTION
**Pull request by petermax2**
## Basics

These points need to be fulfilled for every PR:

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [ ] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

Closes #3806 .

### CODE REVIEW

**Comment by mpranj** [comment]
Thank you for creating the PR! I'll try to save you some CI time:

```
$ ./bin/gen-gpg-testkey
failed to create GPG test key with error message: General error
```

**Comment by petermax2** [comment]
Thank you, @mpranj ! I was expecting that. 

A work-around to fix this issue would be to not create a new key for testing but import an existing one in tool `gen-gpg-testkey`. The unit tests of the gpgme plugin still seem to work. The unit tests import a pre-generated key.

I suggest to import the gpgme testkey instead of creating a new one. What do you think?

**Comment by petermax2** [comment]
> A work-around to fix this issue would be to not create a new key for testing but import an existing one in tool `gen-gpg-testkey`. The unit tests of the gpgme plugin still seem to work. The unit tests import a pre-generated key.

Now I remember why we created a new key in the first place. If we import an existing key, it is not marked as trusted. GPG fails with all following operations, because of the untrusted key. Newly created keys are trusted per default. which mitigates the problem.

**Comment by mpranj** [comment]
I can also reproduce _some_ problem on my Fedora 33 system. I added `gpgme_strsource` and get:
```
failed to create GPG test key with error message: GPG Agent: No such file or directory
```
The error occurs when no `.gnupg` directory existed before the key is generated. It also occurs on subsequent calls.

The error is gone when I use my existing old `.gnupg` directory where everything is set up.

---
If I do this before the tests on Fedora 33, it fixes everything: :eyes: 
```
rm -Rf ~/.gnupg
gpgconf --kill gpg-agent
```
**But** it seems it does not fix the problem on macOS.

**Comment by petermax2** [comment]
gpg-agent is not made for CI testing. We already established that in countless Github issues. :laughing: 

In the prepared Arch VM with GnupG 2.3.1 installed, I still get the error:

    [vagrant@archlinux build]$ rm -rf ~/.gnupg/
    [vagrant@archlinux build]$ gpgconf --kill gpg-agent
    [vagrant@archlinux build]$ ./bin/gen-gpg-testkey 
    failed to create GPG test key with error message: GPGME: General error

I do not get any errors in the Fedora 33 VM tough:

    [vagrant@fedora33 build]$ ./bin/gen-gpg-testkey 
    BC83F82CB70968C1F7E1DB5A0426DEC663EC4C62[vagrant@fedora33 build]$ 
    [vagrant@fedora33 build]$ gpg --version
    gpg (GnuPG) 2.2.25
    libgcrypt 1.8.7

_output omitted_

    [vagrant@fedora33 build]$ rm -rf ~/.gnupg/
    [vagrant@fedora33 build]$ gpgconf --kill gpg-agent
    [vagrant@fedora33 build]$ ./bin/gen-gpg-testkey 
    B252E5E6A302B85D12EBB33B5E85FAFE0BB5D882[vagrant@fedora33 build]$ ./bin/gen-gpg-testkey 
    B252E5E6A302B85D12EBB33B5E85FAFE0BB5D882[vagrant@fedora33 build]$ 
    [vagrant@fedora33 build]$ ./bin/gen-gpg-testkey 
    B252E5E6A302B85D12EBB33B5E85FAFE0BB5D882[vagrant@fedora33 build]$ 
    [vagrant@fedora33 build]$ ./bin/gen-gpg-testkey 
    B252E5E6A302B85D12EBB33B5E85FAFE0BB5D882[vagrant@fedora33 build]$ 
    [vagrant@fedora33 build]$ 

I'm afraid the error lies within gpgme and there is nothing we can do about it at the moment.

**Comment by mpranj** [comment]
> I'm afraid the error lies within gpgme and there is nothing we can do about it at the moment.

I thought that there might have been some breaking change due to the related package upgrades.
If the issue is upstream, I would not waste any time on this.

**Comment by petermax2** [comment]
> I thought that there might have been some breaking change due to the related package upgrades.

Can we maybe suggest to not update GnuPG to 2.3.x yet? I would suggest that we keep an eye on gpgme releases, that might follow in the near future.

**Comment by mpranj** [comment]
> Can we maybe suggest to not update GnuPG to 2.3.x yet? I would suggest that we keep an eye on gpgme releases, that might follow in the near future.

Definitely. Let's do that and I'll try to change the CI and pin to version 2.2.x in the mean time, so that we still run the GPG tests.

**Comment by petermax2** [comment]
@mpranj You were able to reproduce the problem with `gen-gpg-testkey`. I got feedback from the GnuPG developers and tried if it helps. Could you maybe run this version of `gen-gpg-testkey` and paste me the output here?

**Comment by mpranj** [comment]
On 413bdcab3339e87dabd8abaec76d21b0fe44fb72 the `testshell_markdown_tutorial_crypto` tests succeed and `gen-gpg-testkey` works. 🚀 

On 2426269679a878b993a92b5f62e1d0302a0c531a there seems to be a regression due to the passphrase:

```
213: CMD: kdb file user:/tests/password | xargs cat
213: RET: 0
213: STDERR: Get returned with an exception: Sorry, module fcrypt issued the error C01330:@Plugin Misbehavior: GPG failed with return value 512. gpg: encrypted with 3072-bit RSA key, ID BF762303C43B3535, created 2021-05-06@      "elektra testkey (gen-gpg-testkey)"@gpg: public key decryption failed: Inappropriate ioctl for device@gpg: decryption failed: No secret key@@Mountpoint: user:/tests@Configfile: /Users/mpranj/.config/test.ini@At: ../src/plugins/crypto/gpg.c:788@ (in execute at src/tools/kdb/file.cpp:47)
213: STDOUT: -----BEGIN PGP MESSAGE-----
213:
213: hQGMA792IwPEOzU1AQv/RMuk9gHRtINxppHgT3mArd5mrMECKyKEJgTDgtxVJKSK
213: lSUHAi4mjBvS/r70L474KqOWN4cGPWxmP1ftRK2WTLRjcqb8CTzHJ6I/bXorgnz2
213: kVPmF+Vlp9MQNwJVz7TqQWYAc1vy4uoxaRVVsCrjs4wuAl/B0Q3H5rLw7jzfDzuu
213: qOE0d1+DwQbXWCyf09iwQIU/mYHgDWsJG/6t8N/4QVxo+KnBbU4pyDECUUhzgwnk
213: aTflZ8B6Qy/R9kCBw0hS7fpv0mawNLLTK2mWAN5NQQKezjW3Zq63lOC5L4ScH0uT
213: x0xmx8qXIKlF7Lcj0ItYpaWEyjQcZdSL11OdUSzTsyE5DHUjhafucfXuvgkPXJc+
213: fvyATuSre9rXStno6rGz6/a+PqOd8iG3fzAwWOehnpMvS3Kugw6TDo8yIIhFUjT4
213: Gad+It+W+iyM8TZPQeF3rGIe1ksb7O3Mh8/5BVfLkY5kOm9++brfQXlcvtbU9Mbw
213: fdVBX8gl3czy2A94egPR0rAB5HNQs9R1BbWSmIRt48rZb8UrLsH8Uh/xErmV02ul
213: otAyvUV/bpOeQA85h8bRIvoru3t0TLW/ztjiNRjyOFzkcAsuyBkjpDgAlD1ujUax
213: dEMsaBG14TQTox2Y59d1LgRZHWxBK8roW1yx95xrAVLcz00v2NTB0N8I2f275RVR
213: 5gs86W3pLS6x46q0BChK99+20g4DqZ9UFbQVNRwW1zhGVZ7vTVSkVJS+G0yTG0t4
213: lw==
213: =nXuO
213: -----END PGP MESSAGE-----
213: ERROR: C01330@Plugin Misbehavior: GPG failed with return value 512. gpg: encrypted with 3072-bit RSA key, ID BF762303C43B3535, created 2021-05-06@      "elektra testkey (gen-gpg-testkey)"@gpg: public key decryption failed: Inappropriate ioctl for device@gpg: decryption failed: No secret key@@Mountpoint: user:/tests@Configfile: /Users/mpranj/.config/test.ini@At: ../src/plugins/crypto/gpg.c:788@ (in execute at src/tools/kdb/file.cpp:47)
```

**Comment by petermax2** [comment]
> On [413bdca](https://github.com/ElektraInitiative/libelektra/commit/413bdcab3339e87dabd8abaec76d21b0fe44fb72) the `testshell_markdown_tutorial_crypto` tests succeed and `gen-gpg-testkey` works. rocket

but not on the build server?

**Comment by mpranj** [comment]
> but not on the build server?

I'm sorry, I gave misleading information. I didn't notice I had the old GnuPG 2.2.x in my PATH and there it worked. :disappointed: 

With GnuPG 2.3.x I again get a `GPGME: General error` when running `gen-gpg-testkey`.

**Comment by petermax2** [comment]
> EDIT: but the buildserver should have the old GnuPG as well. eyes

It seems like `gpgme` pulls in `gnupg-2.3.1` as dependency.

**Comment by mpranj** [comment]
> It seems like gpgme pulls in gnupg-2.3.1 as dependency.

Yes, but I have overridden this by providing GnuPG 2.2.x in PATH before the newer GnuPG. That change fixed the problems in our CI.

**Comment by petermax2** [comment]
> Yes, but I have overridden this by providing GnuPG 2.2.x in PATH before the newer GnuPG. That change fixed the problems in our CI.

But if you look at the [Cirrus Logs](https://cirrus-ci.com/task/5874860288114688?logs=install#L5689)

    ==> Installing gpgme dependency: gnupg

...

    ==> Downloading https://ghcr.io/v2/homebrew/core/gnupg/manifests/2.3.1

I think `gpgme` depends on current `gnupg`.

**Comment by mpranj** [comment]
> I think gpgme depends on current gnupg.

The dependency is there in homebrew, but it works with 2.2.x as well. As I said, it can be tricked by setting 2.2.x in the PATH before 2.3.x. This is how I fixed master builds. You can rebase or check Cirrus for master builds.

My Fedora 34 machine has gpgme 1.15.1 with gnupg 2.2.27, so I thought they are probably compatible.

In the long run we definitely want to avoid this workaround.

**Comment by petermax2** [comment]
> In the long run we definitely want to avoid this workaround.

Of course! I would also like to fix the issue. The GnuPG developers do not get any problems with the master versions of gnupg and gpgme when running `gen-gpg-testkey` on their systems.

**Comment by mpranj** [comment]
> The GnuPG developers do not get any problems with the master versions of gnupg and gpgme when running gen-gpg-testkey on their systems.

I also could not reproduce the issue at first, due to an existing & intact .gnupg directory. When I moved ~/.gnupg to start fresh (like the CI) the error appeared. That _might_ explain why they have no issues.

Also, I think they ran their own [run-genkey](https://dev.gnupg.org/source/gpgme/browse/master/tests/run-genkey.c$435) according to the [bug report](https://dev.gnupg.org/T5422), and not Elektra's `gen-gpg-testkey`.

**Comment by petermax2** [comment]
> Also, I think they ran their own [run-genkey](https://dev.gnupg.org/source/gpgme/browse/master/tests/run-genkey.c$435) according to the [bug report](https://dev.gnupg.org/T5422), and not Elektra's `gen-gpg-testkey`.

Their `run-genkey` is working without a problem in my prepared ArchVM.

    [vagrant@archlinux gpgme]$ ./tests/run-genkey --loopback "elektra testkey"
    Generated key: 19161EEDBCF09FF141840301787BDF3D7FF23863 (primary, sub)
    [vagrant@archlinux gpgme]$ gpg --version
    gpg (GnuPG) 2.3.2-beta27

Generating unprotected keys also works fine

    [vagrant@archlinux gpgme]$ ./tests/run-genkey --unprotected "elektra test test key"
    Generated key: 961011B6349C68A0BE78C0112DE677D786BA9DCE (primary, sub)

I am now checking for differences between `run-genkey` and `gen-gpg-testkey`.

**Comment by petermax2** [comment]
:rocket: 

    219/231 Test #185: testshell_markdown_tutorial_crypto ............   Passed    5.10 sec

I think the dbus errors are hopefully not caused by this change? 

    The following tests FAILED:
	     54 - testmod_dbus (Failed)
   	     55 - testmod_dbusrecv (Failed)

File: .github/workflows/macOS_gcc10.yml [hunk 1]
```diff
@@ -42,7 +42,7 @@ jobs:
           dbus \
           discount \
           flex \
-          gnupg@2.2 \
+          gnupg \
```
Comment (mpranj) [inline .github/workflows/macOS_gcc10.yml:45]: There is a PATH at line 61, please remove the whole line. (GitHub does not let me comment there 😕)

Also gnupg not needed explicitly here (same as above).

**Review by mpranj — APPROVED** [review]
Thank you so much for fixing it and for maintaining your plugins! The plugins have been really stable in the CI for the past year! ❤️ 

I only left small cleanup notes, if you could do that.

Just out of interest: how did you fix the problem and are we now using a different protocol to communicate with gpgme?

**Comment by mpranj** [comment]
> I think the dbus errors are hopefully not caused by this change? 

No, these errors occur sporadically, but the test should be automatically re-run. If the second run succeeds everything is fine.

**Comment by petermax2** [comment]
> I only left small cleanup notes, if you could do that.

Sure thing! Thank you for the feedback!

> Just out of interest: how did you fix the problem and are we now using a different protocol to communicate with gpgme?

I removed two flags from the call of `gpgme_op_createkey`. These flags (`GPGME_CREATE_SIGN | GPGME_CREATE_ENCR`) seem to cause the error. I reported this insight upstream.

**Comment by petermax2** [comment]
> I removed two flags from the call of `gpgme_op_createkey`. These flags (`GPGME_CREATE_SIGN | GPGME_CREATE_ENCR`) seem to cause the error. I reported this insight upstream.

Here is the explanation by the GnuPG team:

_I guess the problem with the flags is that gnupg 2.2.27 creates an RSA key by default which can be used for signing and encryption, but gnupg 2.3 creates a key with an ed25519 key as primary key and such a key cannot be used for encryption. Explicitly requesting the creation of an RSA key might have fixed the issue for you as well. But it's better to go with the defaults unless you need a specific algorithm._

**Comment by mpranj** [comment]
Thank you again for fixing this! Great work! 🚀 

> Here is the explanation by the GnuPG team:

Thanks also for the investigation and explanation.

**Pull request merged**